src/Core/Material/BlinnPhongMaterialModel.cpp

```diff
@@ -24,6 +26,81 @@ void BlinnPhongMaterialModel::displayInfo() const {
     print( hasNormalTexture(), " Normal Texture : ", m_texNormal );
     print( hasOpacityTexture(), " Alpha Texture  : ", m_texOpacity );
 }
+Utils::Color Material::BlinnPhongMaterialModel::evalBSDF( Vector3f w_i,
+                                                          Vector3f w_o,
+                                                          Vector3f normal,
+                                                          Vector2f uv ) {
+    // diffuse lambertien component
+    Utils::Color diffuse = m_kd / M_PI;
+
+    // Blinn-Phong specular component
+    Vector3f halfway = w_i + w_o;
+    halfway.normalize();
+    Scalar specularIntensity = (m_ns + 2.0f) / (2.0f * M_PI) * std::pow(normal.dot(halfway), m_ns);
+    Utils::Color specular = m_ks * specularIntensity;
+
+    // Combine the diffuse and specular components
+    Utils::Color bsdf = diffuse + specular;
+}
+
+std::optional<std::pair<Vector3f, Scalar>>
+BlinnPhongMaterialModel::sample( Vector3f inDir, Vector3f normal, Vector2f u ) {
+    /*
+    compute intensity of diffuse and specular
+    compute norm (max(1, Idiffuse + Ispecular))
+    normalize Idiffuse and Ispecular
+    */
+
+    Vector3f nextDir, halfway;
+
+    Scalar dIntensity = (m_kd.x() + m_kd.y() + m_kd.z()) / 3.f;
+    Scalar sIntensity = (m_ks.x() + m_ks.y() + m_ks.z()) / 3.f;
+    Scalar diffSpecNorm = std::max(1.f, dIntensity + sIntensity);
+
+    dIntensity /= diffSpecNorm;
+    sIntensity /= diffSpecNorm;
+
+    std::uniform_real_distribution<Scalar> unifDistributionRand{0, 1};
+    Scalar distrib = unifDistributionRand(m_randomEngine);
+
+    // diffuse part
+    if(distrib < dIntensity) {
+        nextDir = sampleHemisphereCosineWeighted(u);
```

Comment:
Your `eval` method seems to take input directions and normal in world frame. But here, your sampling method return the result in local frame. This is not documented and may be difficult to infer. More, it should be helpful to have both.
see `LambertianMaterialModel::sample` where you do the right thing.